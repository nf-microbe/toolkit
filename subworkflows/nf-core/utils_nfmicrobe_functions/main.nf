//
// Subworkflow with functionality that is required for nf-microbe subworkflows
//

/*
========================================================================================
    SUBWORKFLOW DEFINITION
========================================================================================
*/

workflow UTILS_NFMICROBE_FUNCTIONS {

    main:

    ch_dummy_channel = Channel.empty()

    emit:
    dummy_emit = true

}

/*
========================================================================================
    FUNCTIONS
========================================================================================
*/

//
// Identify work directories to clean
//
def getWorkDirs(ch_to_clean, ch_downstream, print_output) {
    // combine channel to clean and dependent channel to clean only channels that have an output
    def ch_double_workdir1 = Channel.empty()
    def ch_double_workdir2 = Channel.empty()
    def ch_branch = ch_to_clean.combine(ch_downstream, by:0)
        .branch { it ->
            double_ch: it.size() > 3
            single_ch: true
        }
    ch_double_workdir1 = ch_branch.double_ch
        .map { meta, ch_to_clean1, ch_to_clean2, ch_dep ->
            return [ [ meta ], ch_to_clean1, ch_dep ]
        }
    ch_double_workdir2 = ch_branch.double_ch
        .map { meta, ch_to_clean1, ch_to_clean2, ch_dep ->
            return [ [ meta ], ch_to_clean2, ch_dep ]
        }

    def ch_workdirs = ch_double_workdir1.mix(ch_double_workdir2).mix(ch_branch.single_ch)
        // filter to retain work directory
        .map { meta, files_to_clean, dependent_files ->
            // do not clean directory if it is not a work directory
            if (( files_to_clean =~ /(^.*\/work\/[^\/]+\/[^\/]+\/).*/ )) {
                def dir_to_clean = ( files_to_clean =~ /(^.*\/work\/[^\/]+\/[^\/]+\/).*/ )[0][1]
                    return [ [ id: meta.id ], dir_to_clean ]
            }
        }
    // remove redundancy
    .unique()

    // print output for tests, otherwise should be false
    if (print_output) {
        ch_workdirs.view()
    }

    return ch_workdirs
}

//
// Filter channels to remove empty fastq files
//
def rmEmptyFastQs(ch_fastqs, print_output) {
    def ch_nonempty_fastqs = ch_fastqs
        .filter { meta, fastq ->
                if ( meta.single_end ) {
                    try {
                        fastq.countFastq(limit: 10) > 1
                    } catch (java.util.zip.ZipException e) {
                        log.warn "[rmEmptyFastQs]: ${fastq} is not in GZIP format, this is likely because it was cleaned with --remove_intermediate_files"
                        true
                    } catch (EOFException) {
                        log.warn "[rmEmptyFastQs]: ${fastq} has an EOFException, this is likely an empty gzipped file."
                    }
                } else {
                    try {
                        fastq[0].countLines( limit: 10 ) > 1
                    } catch (java.util.zip.ZipException e) {
                        log.warn "[rmEmptyFastQs]: ${fastq} is not in GZIP format, this is likely because it was cleaned with --remove_intermediate_files"
                        true
                    } catch (EOFException) {
                        log.warn "[rmEmptyFastQs]: ${fastq[0]} has an EOFException, this is likely an empty gzipped file."
                    }
                }
            }

    if (print_output) {
        ch_nonempty_fastqs.view()
    }

    return ch_nonempty_fastqs
}

