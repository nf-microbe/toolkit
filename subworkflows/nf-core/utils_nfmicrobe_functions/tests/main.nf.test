nextflow_function {

    name "Test nf-microbe functions"
    script "../main.nf"

    tag 'subworkflows'
    tag "subworkflows_nfcore"
    tag 'utils_nfmicrobe_functions'
    tag 'subworkflows/utils_nfmicrobe_functions'

    test("getWorkDirs: single workdir") {
        function "getWorkDirs"
        when {
            function {
            """
            input[0] = Channel.of(
                [
                    [ id:'test' ], // meta map
                    "/somepath/work/12/hashforworkdirtoclean/somefile.txt"
                ]
            )
            input[1] = Channel.of(
                [
                    [ id:'test' ], // meta map
                    "/somepath/work/34/hashforworkdirdownstream/someotherfile.txt"
                ]
            )
            input[2] = true
            """
            }
        }
        then {
            assertAll (
                { assert function.success },
                { assert function.stdout.any() { it.contains("[[id:test], /somepath/work/12/hashforworkdirtoclean/]") } },
                { assert function.stdout.any() { !it.contains("/somepath/work/34/hashforworkdirdownstream/") } }
            )
        }
    }

    test("getWorkDirs: two workdirs") {
        function "getWorkDirs"
        when {
            function {
            """
            input[0] = Channel.of(
                [
                    [ id:'test' ], // meta map
                    "/somepath/work/12/hashforworkdirtoclean/somefile.txt"
                ],
                [
                    [ id:'test' ], // meta map
                    "/somepath/work/24/hashforworkdirtoclean/somefile.txt"
                ]
            )
            input[1] = Channel.of(
                [
                    [ id:'test' ], // meta map
                    "/somepath/work/34/hashforworkdirdownstream/someotherfile.txt"
                ]
            )
            input[2] = true
            """
            }
        }
        then {
            assertAll (
                { assert function.success },
                { assert function.stdout.any() { it.contains("[[id:test], /somepath/work/12/hashforworkdirtoclean/]") } },
                { assert function.stdout.any() { it.contains("[[id:test], /somepath/work/24/hashforworkdirtoclean/]") } },
                { assert function.stdout.any() { !it.contains("/somepath/work/34/hashforworkdirdownstream/") } }
            )
        }
    }

    test("rmEmptyFastQs: paired end + cleaned.fasta.gz") {
        function "rmEmptyFastQs"
        when {
            function {
            """
            input[0] = Channel.of(
                [
                    [ id:'nonempty', single_end:false], // meta map
                    [
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkifExists: true),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkifExists: true)
                    ]
                ],
                [
                    [ id:'cleaned', single_end:true ], // meta map
                    file("${projectDir}/assets/test-datasets/genomics/sarscov2/illumina/fastq/cleaned.fastq.gz", checkIfExists: true)
                ]
            )
            input[1] = true
            """
            }
        }
        then {
            assertAll (
                { assert function.success },
                { assert function.stdout.any() { it.contains("[[id:nonempty, single_end:false], [/nf-core/test-datasets/modules/data/genomics/sarscov2/illumina/fastq/test_1.fastq.gz, /nf-core/test-datasets/modules/data/genomics/sarscov2/illumina/fastq/test_2.fastq.gz]]") } },
                { assert function.stdout.any() { it.contains("[[id:cleaned, single_end:true]") } }
            )
        }
    }

    test("rmEmptyFastQs: single end + empty.fastq.gz") {
        function "rmEmptyFastQs"
        when {
            function {
            """
            input[0] = Channel.of(
                [
                    [ id:'nonempty', single_end: true], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkifExists: true)
                ],
                [
                    [ id:'empty', single_end:true ], // meta map
                    file("${projectDir}/assets/test-datasets/genomics/sarscov2/illumina/fastq/empty.fastq.gz", checkIfExists: true)
                ]
            )
            input[1] = true
            """
            }
        }
        then {
            assertAll (
                { assert function.success },
                { assert function.stdout.any() { it.contains("[[id:nonempty, single_end:true], /nf-core/test-datasets/modules/data/genomics/sarscov2/illumina/fastq/test_1.fastq.gz]") } },
                { assert function.stdout.any() { !it.contains("id:empty") } }
            )
        }
    }
}