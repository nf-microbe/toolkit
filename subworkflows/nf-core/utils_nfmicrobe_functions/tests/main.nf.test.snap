{
    "rmEmptyFastQs: single end + empty.fastq.gz": {
        "content": [
            [
                "[[id:nonempty, single_end:true], /nf-core/test-datasets/modules/data/genomics/sarscov2/illumina/fastq/test_1.fastq.gz]"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-27T12:33:33.88962521"
    },
    "getWorkDirs: single workdir": {
        "content": [
            [
                "[[id:test], /somepath/work/12/hashforworkdirtoclean/]"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-27T12:33:18.337800595"
    },
    "rmEmptyFastQs: paired end + cleaned.fasta.gz": {
        "content": [
            "[[id:nonempty, single_end:false], [/nf-core/test-datasets/modules/data/genomics/sarscov2/illumina/fastq/test_1.fastq.gz, /nf-core/test-datasets/modules/data/genomics/sarscov2/illumina/fastq/test_2.fastq.gz]]",
            "[[id:cleaned, single_end:true],"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-27T12:33:28.729491491"
    },
    "rmEmptyFastQs: single end + empty.fastq.gz - stub": {
        "content": [
            "[[id:nonempty, single_end:true], /nf-core/test-datasets/modules/data/genomics/sarscov2/illumina/fastq/test_1.fastq.gz]",
            "[[id:empty, single_end:true],"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-27T12:33:39.23215877"
    },
    "rmEmptyFastAs: nonempty + empty - stub": {
        "content": [
            "[[id:nonempty], /nf-core/test-datasets/modules/data/genomics/sarscov2/genome/genome.fasta.gz]",
            "[[id:empty],"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-27T12:33:55.712194713"
    },
    "rmEmptyFastAs: nonempty + cleaned": {
        "content": [
            "[[id:nonempty], /nf-core/test-datasets/modules/data/genomics/sarscov2/genome/genome.fasta.gz]",
            "[[id:cleaned],"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-27T12:38:31.273126602"
    },
    "getWorkDirs: two workdirs": {
        "content": [
            [
                "[[id:test], /somepath/work/12/hashforworkdirtoclean/]",
                "[[id:test], /somepath/work/24/hashforworkdirtoclean/]"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-27T12:33:23.37021052"
    },
    "rmEmptyFastAs: nonempty + empty": {
        "content": [
            [
                "[[id:nonempty], /nf-core/test-datasets/modules/data/genomics/sarscov2/genome/genome.fasta.gz]"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-27T12:33:50.366840648"
    }
}